Made several route line related calls asynchronous with suspend and c…

cafesilencio opened the pull request:
### Description
Resolves #4091 by adding suspend and callback versions to methods that are resource intensive. Also includes some minor performance improvements.

### Changelog
```
<changelog>Added suspend and callback versions of some methods in the MapboxRouteLineApi class.</changelog>
```

cafesilencio commented:
> we might want to do the parsing on a worker thread instead of the main thread

I thought about this and for the suspend function I didn't impose a specific thread pool because the developer can control that when calling the method.  However, for the callback version that's wrapping the suspend call it makes sense. Also, the callback version can be made cancellable.

cafesilencio commented:
I made the callback version cancelable. I also incorporated the fix and test from #4110.  There's some issue with the instrumentation tests with inflating the view. Both myself and Ahbishek are getting an error when running SanityUiRouteTest and the others.  I had to update the other tests so they'd compile but I have no idea if they pass because of the previously mentioned error.  For the time being I added an Ignore to the AlternativeRouteSelectionTest in order to move things forward. I can revisit these tests once we resolve the view inflating error.

Guardiola31337 commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/AlternativeRouteSelectionTest.kt:
```diff
@@ -81,6 +86,7 @@ class AlternativeRouteSelectionTest : BaseTest<BasicNavigationViewActivity>(
         }
     }
 
+    @Ignore
```
@cafesilencio Could this https://github.com/mapbox/mapbox-navigation-android/pull/4089 be related?

cafesilencio replied:
It could be. I revert that change on a branch and see what the result is.

cafesilencio replied:
It also happens with SanityUIRouteTest.

cafesilencio replied:
It fails at BasicNavigationViewActivity -> onCreate -> binding = ActivityBasicNavigationViewBinding...

I think this is a bigger issue outside the scope of this PR.  I tried a quick and dirty fix to remove the binding related calls and set the content view the old fashioned way using the technique we were doing earlier in our test activities where the view layout had a layout container and we were adding the mapview during onCreate and this was also failing. Further investigation is needed.

LukasPaczos replied:
I'm able to run the test and have no build issues, but it's failing:

```
java.lang.AssertionError: expected:<DirectionsRoute{routeIndex=1, distance=1872.7, duration=261.6, durationTypical=null, geometry=mvl`gAnalshFa@iVUuMG}D[{Rq@y`@ScLaB{aA{@gi@[kQGmEIqE[yQ{@uh@aAgl@_@{Ta@qU_Agk@aB{aA{@_j@UmNKoGMoGS{Mm@i^GyEGqDysBhGkg@~AaZ|@kPh@}AD{[bAonB`GeDJmDJkzArEqRl@{FPR|K\jT`@vVJ|GTlN, weight=387.3, weightName=routability, legs=[RouteLeg{distance=1872.7, duration=261.6, durationTypical=null, summary=Fulton Street, 24th Avenue, admins=null, steps=[LegStep{distance=911.0, duration=99.4, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=mvl`gAnalshFa@iVUuMG}D[{Rq@y`@ScLaB{aA{@gi@[kQGmEIqE[yQ{@uh@aAgl@_@{Ta@qU_Agk@aB{aA{@_j@UmNKoGMoGS{Mm@i^GyEGqD, name=Fulton Street, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.49348, 37.772151], bearingBefore=0.0, bearingAfter=87.0, instruction=Head east on Fulton Street, type=depart, modifier=null, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=911.0, announcement=Head east on Fulton Street, then turn left onto 24th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Head east on Fulton Street, then turn left onto <say-as interpret-as="address">24th</say-as> Avenue</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=137.5, announcement=Turn left onto 24th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Turn left onto <say-as interpret-as="address">24th</say-as> Avenue</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=911.0, primary=BannerText{text=24th Avenue, components=[BannerComponents{text=24th Avenue, type=text, subType=null, abbreviation=24th Ave, abbreviationPriority=0, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=turn, modifier=left, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=156.2, intersections=[StepIntersection{rawLocation=[-122.49348, 37.772151], bearings=[87], classes=null, entry=[true], in=null, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.492777, 37.772183], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.491708, 37.772232], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.490638, 37.772281], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.489565, 37.772329], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, true, false], in=3, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.488492, 37.772378], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.487418, 37.772427], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.486349, 37.772476], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.485279, 37.772525], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.484208, 37.772572], bearings=[0, 90, 150, 210, 270], classes=null, entry=[false, true, false, true, false], in=4, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=831.5, duration=129.7, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=wsm`gA|zwrhFysBhGkg@~AaZ|@kPh@}AD{[bAonB`GeDJmDJkzArEqRl@, name=24th Avenue, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.483135, 37.77262], bearingBefore=85.0, bearingAfter=355.0, instruction=Turn left onto 24th Avenue, type=turn, modifier=left, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=811.5, announcement=Continue on 24th Avenue for a half mile, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Continue on <say-as interpret-as="address">24th</say-as> Avenue for a half mile</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=448.8, announcement=In a quarter mile, turn left onto Geary Boulevard, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">In a quarter mile, turn left onto Geary Boulevard</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=96.2, announcement=Turn left onto Geary Boulevard, then you will arrive at your destination, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Turn left onto Geary Boulevard, then you will arrive at your destination</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=831.5, primary=BannerText{text=Geary Boulevard, components=[BannerComponents{text=Geary Boulevard, type=text, subType=null, abbreviation=Geary Blvd, abbreviationPriority=0, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=turn, modifier=left, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=141.9, intersections=[StepIntersection{rawLocation=[-122.483135, 37.77262], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.483268, 37.774489], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, true], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.483405, 37.776355], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, true], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.48354, 37.778222], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, true], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=130.2, duration=32.5, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=gf|`gAt|xrhF{FPR|K\jT`@vVJ|GTlN, name=Geary Boulevard, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.483675, 37.780084], bearingBefore=355.0, bearingAfter=355.0, instruction=Turn left onto Geary Boulevard, type=turn, modifier=left, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=20.0, announcement=You have arrived at your destination, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">You have arrived at your destination</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=130.2, primary=BannerText{text=You will arrive, components=[BannerComponents{text=You will arrive, type=text, subType=null, abbreviation=null, abbreviationPriority=null, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=arrive, modifier=straight, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}, BannerInstructions{distanceAlongGeometry=20.0, primary=BannerText{text=You have arrived, components=[BannerComponents{text=You have arrived, type=text, subType=null, abbreviation=null, abbreviationPriority=null, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=arrive, modifier=straight, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=89.2, intersections=[StepIntersection{rawLocation=[-122.483675, 37.780084], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, false], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.483684, 37.78021], bearings=[0, 90, 180, 270], classes=null, entry=[true, false, false, true], in=2, out=3, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.484756, 37.780162], bearings=[0, 90, 180, 270], classes=null, entry=[true, false, false, true], in=1, out=3, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=0.0, duration=0.0, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=mj|`gAto{rhF, name=Geary Boulevard, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.485003, 37.780151], bearingBefore=267.0, bearingAfter=0.0, instruction=You have arrived at your destination, type=arrive, modifier=null, exit=null}, voiceInstructions=[], bannerInstructions=[], drivingSide=right, weight=0.0, intersections=[StepIntersection{rawLocation=[-122.485003, 37.780151], bearings=[87], classes=null, entry=[true], in=0, out=null, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}], incidents=null, annotation=LegAnnotation{distance=null, duration=null, speed=null, maxspeed=[MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}], congestion=null}}], routeOptions=RouteOptions{baseUrl=https://api.mapbox.com, user=mapbox, profile=driving-traffic, coordinates=[Point{type=Point, bbox=null, coordinates=[-122.4934801, 37.7721532]}, Point{type=Point, bbox=null, coordinates=[-122.4850055, 37.7801765]}], alternatives=true, language=en, radiuses=null, bearings=null, continueStraight=false, roundaboutExits=false, geometries=polyline6, overview=full, steps=true, annotations=maxspeed, exclude=null, voiceInstructions=true, bannerInstructions=true, voiceUnits=imperial, accessToken=tokenOmitted, requestUuid=ckl8j5te2ch2i79r7cmgw67sj, approaches=null, waypointIndices=null, waypointNames=null, waypointTargets=null, walkingOptions=null}, voiceLanguage=en-US}> but was:<DirectionsRoute{routeIndex=0, distance=1683.8, duration=207.3, durationTypical=null, geometry=mvl`gAnalshFa@iVUuMG}D[{Rq@y`@ScLaB{aA{@gi@[kQGmEIqE[yQ{@uh@aAgl@_@{Ta@qU_Agk@wl@hBiRj@eELag@|AiDJkELimB~FusBlGwAg|@IuDGwDoAwu@GmEsDLcIVooAxDqCHoFPaGP}FPTlN, weight=365.4, weightName=routability, legs=[RouteLeg{distance=1683.8, duration=207.3, durationTypical=null, summary=Fulton Street, 27th Avenue, admins=null, steps=[LegStep{distance=628.0, duration=63.099999999999994, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=mvl`gAnalshFa@iVUuMG}D[{Rq@y`@ScLaB{aA{@gi@[kQGmEIqE[yQ{@uh@aAgl@_@{Ta@qU_Agk@, name=Fulton Street, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.49348, 37.772151], bearingBefore=0.0, bearingAfter=87.0, instruction=Head east on Fulton Street, type=depart, modifier=null, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=628.0, announcement=Head east on Fulton Street, then turn left onto 27th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Head east on Fulton Street, then turn left onto <say-as interpret-as="address">27th</say-as> Avenue</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=149.3, announcement=Turn left onto 27th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Turn left onto <say-as interpret-as="address">27th</say-as> Avenue</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=628.0, primary=BannerText{text=27th Avenue, components=[BannerComponents{text=27th Avenue, type=text, subType=null, abbreviation=27th Ave, abbreviationPriority=0, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=turn, modifier=left, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=107.7, intersections=[StepIntersection{rawLocation=[-122.49348, 37.772151], bearings=[87], classes=null, entry=[true], in=null, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.492777, 37.772183], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.491708, 37.772232], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.490638, 37.772281], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.489565, 37.772329], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, true, false], in=3, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.488492, 37.772378], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.487418, 37.772427], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=623.9, duration=72.3, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=wjm`gAxc~rhFwl@hBiRj@eELag@|AiDJkELimB~FusBlG, name=27th Avenue, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.486349, 37.772476], bearingBefore=85.0, bearingAfter=355.0, instruction=Turn left onto 27th Avenue, type=turn, modifier=left, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=603.9, announcement=In a half mile, turn right onto Anza Street, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">In a half mile, turn right onto Anza Street</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=129.4, announcement=Turn right onto Anza Street, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Turn right onto Anza Street</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=623.9, primary=BannerText{text=Anza Street, components=[BannerComponents{text=Anza Street, type=text, subType=null, abbreviation=Anza St, abbreviationPriority=0, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=turn, modifier=right, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=84.9, intersections=[StepIntersection{rawLocation=[-122.486349, 37.772476], bearings=[0, 90, 270], classes=null, entry=[true, true, false], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.486484, 37.774342], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, true], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.486619, 37.776209], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, true], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=188.6, duration=33.1, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=whx`gAb}~rhFwAg|@IuDGwDoAwu@GmE, name=Anza Street, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.486754, 37.778076], bearingBefore=355.0, bearingAfter=85.0, instruction=Turn right onto Anza Street, type=turn, modifier=right, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=168.6, announcement=In 600 feet, turn left onto 25th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">In 600 feet, turn left onto <say-as interpret-as="address">25th</say-as> Avenue</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=85.5, announcement=Turn left onto 25th Avenue, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Turn left onto <say-as interpret-as="address">25th</say-as> Avenue</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=188.6, primary=BannerText{text=25th Avenue, components=[BannerComponents{text=25th Avenue, type=text, subType=null, abbreviation=25th Ave, abbreviationPriority=0, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=turn, modifier=left, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=77.6, intersections=[StepIntersection{rawLocation=[-122.486754, 37.778076], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, true], in=2, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.485683, 37.778125], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, true, false], in=3, out=1, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=207.4, duration=23.5, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=ynx`gAfwzrhFsDLcIVooAxDqCHoFPaGP, name=25th Avenue, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.484612, 37.778173], bearingBefore=85.0, bearingAfter=355.0, instruction=Turn left onto 25th Avenue, type=turn, modifier=left, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=187.4, announcement=In 700 feet, turn left onto Geary Boulevard, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">In 700 feet, turn left onto Geary Boulevard</prosody></amazon:effect></speak>}, VoiceInstructions{distanceAlongGeometry=132.4, announcement=Turn left onto Geary Boulevard, then you will arrive at your destination, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">Turn left onto Geary Boulevard, then you will arrive at your destination</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=207.4, primary=BannerText{text=Geary Boulevard, components=[BannerComponents{text=Geary Boulevard, type=text, subType=null, abbreviation=Geary Blvd, abbreviationPriority=0, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=turn, modifier=left, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=35.5, intersections=[StepIntersection{rawLocation=[-122.484612, 37.778173], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, true, false], in=3, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=35.9, duration=15.3, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=ec|`gAt_{rhF}FPTlN, name=Geary Boulevard, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.484747, 37.780035], bearingBefore=355.0, bearingAfter=355.0, instruction=Turn left onto Geary Boulevard, type=turn, modifier=left, exit=null}, voiceInstructions=[VoiceInstructions{distanceAlongGeometry=11.7, announcement=You have arrived at your destination, ssmlAnnouncement=<speak><amazon:effect name="drc"><prosody rate="1.08">You have arrived at your destination</prosody></amazon:effect></speak>}], bannerInstructions=[BannerInstructions{distanceAlongGeometry=35.9, primary=BannerText{text=You will arrive, components=[BannerComponents{text=You will arrive, type=text, subType=null, abbreviation=null, abbreviationPriority=null, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=arrive, modifier=straight, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}, BannerInstructions{distanceAlongGeometry=11.7, primary=BannerText{text=You have arrived, components=[BannerComponents{text=You have arrived, type=text, subType=null, abbreviation=null, abbreviationPriority=null, imageBaseUrl=null, imageUrl=null, directions=null, active=null}], type=arrive, modifier=straight, degrees=null, drivingSide=null}, secondary=null, sub=null, view=null}], drivingSide=right, weight=59.7, intersections=[StepIntersection{rawLocation=[-122.484747, 37.780035], bearings=[0, 90, 180, 270], classes=null, entry=[true, true, false, false], in=2, out=0, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}, StepIntersection{rawLocation=[-122.484756, 37.780162], bearings=[0, 90, 180, 270], classes=null, entry=[true, false, false, true], in=2, out=3, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}, LegStep{distance=0.0, duration=0.0, durationTypical=null, speedLimitUnit=mph, speedLimitSign=mutcd, geometry=mj|`gAto{rhF, name=Geary Boulevard, ref=null, destinations=null, mode=driving, pronunciation=null, rotaryName=null, rotaryPronunciation=null, maneuver=StepManeuver{rawLocation=[-122.485003, 37.780151], bearingBefore=267.0, bearingAfter=0.0, instruction=You have arrived at your destination, type=arrive, modifier=null, exit=null}, voiceInstructions=[], bannerInstructions=[], drivingSide=right, weight=0.0, intersections=[StepIntersection{rawLocation=[-122.485003, 37.780151], bearings=[87], classes=null, entry=[true], in=0, out=null, lanes=null, geometryIndex=null, isUrban=null, adminIndex=null, restStop=null, tollCollection=null, mapboxStreetsV8=null, tunnelName=null}], exits=null}], incidents=null, annotation=LegAnnotation{distance=null, duration=null, speed=null, maxspeed=[MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=56, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=null, unit=null, unknown=true, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}, MaxSpeed{speed=40, unit=km/h, unknown=null, none=null}], congestion=null}}], routeOptions=RouteOptions{baseUrl=https://api.mapbox.com, user=mapbox, profile=driving-traffic, coordinates=[Point{type=Point, bbox=null, coordinates=[-122.4934801, 37.7721532]}, Point{type=Point, bbox=null, coordinates=[-122.4850055, 37.7801765]}], alternatives=true, language=en, radiuses=null, bearings=null, continueStraight=false, roundaboutExits=false, geometries=polyline6, overview=full, steps=true, annotations=maxspeed, exclude=null, voiceInstructions=true, bannerInstructions=true, voiceUnits=imperial, accessToken=tokenOmitted, requestUuid=ckl8j5te2ch2i79r7cmgw67sj, approaches=null, waypointIndices=null, waypointNames=null, waypointTargets=null, walkingOptions=null}, voiceLanguage=en-US}>
at org.junit.Assert.fail(Assert.java:88)
at org.junit.Assert.failNotEquals(Assert.java:834)
at org.junit.Assert.assertEquals(Assert.java:118)
at org.junit.Assert.assertEquals(Assert.java:144)
at com.mapbox.navigation.instrumentation_tests.ui.routeline.AlternativeRouteSelectionTest$selectAlternateRoute$1$2.onRouteProgressChanged(AlternativeRouteSelectionTest.kt:116)
at com.mapbox.navigation.core.trip.session.MapboxTripSession.updateRouteProgress(MapboxTripSession.kt:537)
at com.mapbox.navigation.core.trip.session.MapboxTripSession.access$updateRouteProgress(MapboxTripSession.kt:50)
at com.mapbox.navigation.core.trip.session.MapboxTripSession$updateDataFromNavigatorStatus$updateNavigatorStatusDataJob$1.invokeSuspend(MapboxTripSession.kt:507)
at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
at android.os.Handler.handleCallback(Handler.java:938)
at android.os.Handler.dispatchMessage(Handler.java:99)
at androidx.test.espresso.base.Interrogator.loopAndInterrogate(Interrogator.java:148)
at androidx.test.espresso.base.UiControllerImpl.loopUntil(UiControllerImpl.java:519)
at androidx.test.espresso.base.UiControllerImpl.loopMainThreadUntilIdle(UiControllerImpl.java:441)
at androidx.test.espresso.Espresso$1.run(Espresso.java:277)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at android.os.Handler.handleCallback(Handler.java:938)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:223)
at android.app.ActivityThread.main(ActivityThread.java:7656)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

cafesilencio replied:
I know, SanityUIRouteTest also fails.  It's not a build problem there is some other issue with the instrumentation tests outside of the work done in this PR.  Check out the tests in the main branch. I only noticed it because i had to adjust those tests due to the API changes that resulted from the work in this PR.

LukasPaczos replied:
> It's not a build problem there is some other issue with the instrumentation tests outside of the work done in this PR. Check out the tests in the main branch.

All of the tests on the `main` branch are passing on CI and for me locally, see https://app.circleci.com/pipelines/github/mapbox/mapbox-navigation-android/10809/workflows/b38925bf-4997-4f7d-966c-3c3347d90466/jobs/42470.

> SanityUIRouteTest also fails

This one passes for me, the issue only seems to be with `selectAlternateRoute`.

I think we shouldn't move forward with ignoring this test, it reliably fails and has to highlight some kind of an issue with the PR.

cafesilencio replied:
The tests fail for me locally on the main branch.

android.view.InflateException: Binary XML file line #7 in com.mapbox.navigation.instrumentation_tests:layout/activity_basic_navigation_view: Binary XML file line #7 in com.mapbox.navigation.instrumentation_tests:layout/activity_basic_navigation_view: Error inflating class com.mapbox.maps.MapView
Caused by: android.view.InflateException: Binary XML file line #7 in com.mapbox.navigation.instrumentation_tests:layout/activity_basic_navigation_view: Error inflating class com.mapbox.maps.MapView
Caused by: java.lang.reflect.InvocationTargetException


After further discussion perhaps this is an issue with running the tests on the emulator.

LukasPaczos replied:
Any updates here @cafesilencio?

cafesilencio replied:
The main tests pass on my device so I"ve been working on the route selection test. It's challenging because two of the API calls are now asynchronous.

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt:
```diff
@@ -172,13 +177,31 @@ class MapboxRouteLineApi(
+        route: DirectionsRoute,
+        consumer: MapboxNavigationConsumer<Expected<RouteSetValue, RouteLineError>>
+    ) {
+        ThreadController.getMainScopeAndRootJob().scope.launch {
```
This uses a different job than `setRoutes` so we can run into race conditions where calling `setRoutes` would not actually apply the changes because the result would be overwritten by `updateToPrimaryRoute`.

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt:
```diff
@@ -172,13 +177,31 @@ class MapboxRouteLineApi(
-    fun updateToPrimaryRoute(route: DirectionsRoute): Expected<RouteSetValue, RouteLineError> {
-        val newRoutes = directionsRoutes.filter { it != route }.toMutableList().also {
-            it.add(0, route)
+    suspend fun updateToPrimaryRoute(route: DirectionsRoute):
```
Could you move the `suspend` functions to a file in the internal package (refs https://github.com/mapbox/mapbox-navigation-android/issues/4126)?

cafesilencio replied:
A customer's request for a suspend function was the motivation for this work. Moving the suspend function to an internal package implies it's not accessible to external developers. I"ll move it to an external file but not in an internal package for now and let's discuss further.

LukasPaczos commented on test-app/src/main/java/com/mapbox/navigation/examples/core/camera/MapboxCameraAnimationsActivity.kt:
```diff
@@ -177,9 +179,12 @@ class MapboxCameraAnimationsActivity :
-                routeLineAPI?.setRoutes(listOf(RouteLine(routes[0], null)))?.apply {
-                    ifNonNull(routeLineView, mapboxMap.getStyle()) { view, style ->
-                        view.renderRouteDrawData(style, this)
+                // fixme SBNOTE: this is temporary for compilation, don't keep this
```
good to refactor?

cafesilencio replied:
yeah, missed this one

cafesilencio commented:
@NikitaParakhovnik The test to look at is AlternativeRouteSelectionTest.  There are a couple of calls to MapboxRouteLineApi that are now asynchronous. The class methods have callbacks or there are extension functions in MapboxRouteLineApiExtensions that are suspend functions.  I don't know which will be easier to work with in AlternativeRouteSelectionTest.

korshaknn commented:
@cafesilencio I've pushed changes, test is fixed, please check on your side
I've added `jobs` to `setRoute`/`updateToPrimaryRoute`  to wait until previous operation is finished. Do we need it? Or just cancel current operation if exists?
cc @Guardiola31337 @LukasPaczos

cafesilencio commented:
> I've added jobs to setRoute/updateToPrimaryRoute to wait until previous operation is finished. Do we need it? Or just cancel current operation if exists?

I don't think there's any reason to wait. If the developer has called those methods before they have finished then they aren't going to render the results anyway. The existing jobs should be canceled.

cafesilencio commented:
Looking at line 95 of AlternativeRouteSelectionTest, that's the most important assertion. If I add assertTrue(false) to that line the test still passes.  It seems the test is finishing before that line is reached.

cafesilencio commented:
@korshaknn Selecting an alternative route is crashing for me.  If I checkout my last commit 995c1b0935f3187548de53ee2da2f2c96efc6e5d I don't get a crash when selecting an alternative route but with the latest code it crashes.  

Also, setRoutes() should not be on the main thread.

korshaknn commented:
@cafesilencio before my changes `setRoutes` had the next flow:
 
setRoutes (io dispatcher)  -> setNewRouteData. -> buildDrawRoutesState (main dispatcher)
the most consuming fun `buildDrawRoutesState` was called on the main thread
and result returned on `io` thread

now:
setRoutes (main dispatcher)  -> setNewRouteData. -> buildDrawRoutesState (main dispatcher)
we have all hard operations on the worker thread and result on the main

cafesilencio commented:
You're right that Dispatchers.Main on line 597 wasn't supposed to be there.

cafesilencio commented:
@korshaknn Thanks for taking the time to look into this.  I found some time friday to work on the failing test.  Since a crash was introduced when selecting an alternative route line I decided to revert to my previous commit. I was able to fix the failing espresso test and I made sure that buildDrawRoutesState is not running on the main thread. Thanks for pointing out that mistake.

korshaknn commented:
@cafesilencio I've refactored `MapboxRouteLineView`
- removed `jobControl`, no need to use it, there are no coroutines
- imported some vals from `RouteConstants` and `RouteLayerConstants`, now methods are more compact and readable

there are some questions related to `MapboxRouteLineApi`:
- in `setRoutes()` you use `jobControl.scope.launch(ThreadController.IODispatcher)`
I think we should not use `mainJobControl` to launch background coroutines, also if you do it, the result ` consumer.accept(routeData)` will be on `IO` thread. 
so, it should be `jobControl.scope.launch { ... }` 
I don't see any issues here, there are no consuming operations before `buildDrawRoutesState` call (and this call will be on `IO` thread). BTW, `fun updateToPrimaryRoute()` has even more operations before `buildDrawRoutesState()` and it is launched on `main` dispatcher. As an option you can switch to `IO` inside `setNewRouteData() = withContext(IO)` but it's not a blocker.
- switch to `IO` inside `findClosestRoute()`, now it is launched on main

cafesilencio commented:
>  also if you do it, the result consumer.accept(routeData) will be on IO thread.

I don't think this is a problem. After the data is received the developer would pass it to render() in the view class which will be on the main thread.  The feature is working in the test activity.

cafesilencio commented:
> removed jobControl, no need to use it, there are no coroutines

The purpose of this was to make the call non-blocking.  There's no need for the developer to wait for the rendering calls to complete. The purpose of this work was to improve the performance of the process of selecting an alternative route.

cafesilencio commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt:
```diff
@@ -321,12 +231,18 @@ class MapboxRouteLineView(var options: MapboxRouteLineOptions) {
+        hideLayer(style, RouteLayerConstants.WAYPOINT_LAYER_ID)
+    }
+
+    private fun showLayer(style: Style, layerId: String) {
```
The show and hide methods you added are almost entirely duplicated code. That's the reason there was a single method with a visibility parameter so that there wouldn't be duplicated code.

korshaknn replied:
what about single responsibility?
it's easier to read code with `show/hide` methods than with `updateVisibility`.
and no need to pass Visible param

cafesilencio replied:
The update visibility method had the single responsibility of changing the visibility.  That's all it did. The only difference between the two methods you created is the visibility parameter. DRY.

korshaknn commented:
> The purpose of this was to make the call non-blocking. There's no need for the developer to wait for the rendering calls to complete. The purpose of this work was to improve the performance of the process of selecting an alternative route.

please add docs to make it clear

korshaknn commented:
> I don't think this is a problem. After the data is received the developer would pass it to render() in the view class which will be on the main thread. The feature is working in the test activity.

it's confusing that 2 methods of the same class return results on different threads. We use main/IO in both cases.

korshaknn commented:
what I want to discuss before we merge the PR:
1. I think we should  launch and return result on the `main` thread and switch to a `worker` thread in case of consuming calculations.
we have it in `updateToPrimaryRoute()`, but `setRoutes()` launches and returns the result on a `worker` thread. 

2. add unit tests to check that a new coroutine cancels active one (for all methods). 

3. there was a `jobController` in `MapboxRouteLineView`, it was used to make the call non-blocking.
I've removed this logic, but maybe @cafesilencio wants to return it? I've been thinking about such approach, looks like we might have issues if a customer calls functions one by one. Inner coroutines will have a race and final UI will be changed by both coroutines and will be invalid.

cc @LukasPaczos @Guardiola31337

zugaldia commented:
Hey team - when can we expect this PR to land? It's been reported as a P0 by a customer.

LukasPaczos commented:
Agreeing with @korshaknn points and also wanted to re-iterate comments from https://github.com/mapbox/mapbox-navigation-android/pull/4106#issuecomment-794041680, especially about accessing the variables.

For example, `directionsRoutes` field is currently accessed from `updateToPrimaryRoute` (runs on the main thread), `setRoutes` (runs on a worker thread), `getRouteDrawData` (runs on the main thread), `clearRouteLine` (wipes the state synchonously) and other places. All of those invocations not only are async, but are even called from different threads. This can lead to various race conditions or even concurrent modification exceptions. This is only one example but this happens across the class.

I think we still need iterations over the approach taken in this PR:
- how and where we are storing the state?
- how we are accessing that state?
- should we work on a shared state when executing the long-running operations, or should we use copies of objects? how would memory consumption be impacted?
- if we use copies, how do we synchronize the results that we're trying to save to a shared state?
- what is the callback behavior for competing requests? if we call `updateToPrimaryRoute` immediately followed by `setRoutes`, should the `updateToPrimaryRoute` immediately return with an error, since the state is shared and a new request was dispatched?

It feels to me like just offloading the methods to a worker thread/coroutine will not be enough here. We'll need to reconsider what pieces of the codebase are responsible for generating the data, what pieces are storing it, and how the state access and callback system are managed.

Also, we need to take into account that `ThreadController.IODispatcher` is a thread pool. This means that if we do:
1. call `setRoutes` with a  very **long** route
2. wait a small amount of time
3. call `setRoutes` with a very **short** route

Then most likely, the callbacks will be called first with a short route, and then with a long route, and probably a state will contain the short route and not the long one. This all is going to result in the incorrect route being drawn.

cafesilencio commented:
Still testing some modifications...

cafesilencio commented:
I made several modifications.  All of the async calls run on the main thread with a thread from the IO pool being used for specific data calculation tasks which don't mutate any state. 

Based on this kotlin documentation: https://kotlinlang.org/docs/shared-mutable-state-and-concurrency.html#mutual-exclusion, I added a mutex for calls that are mutating shared state.  I've added instrumentation tests that first sets a long route followed by a short route (with a brief delay). The calls are received sequentially even though the short route call is much faster.  I've also tested manually with a test activity and the callback always receives the result of the first call followed by the result of the second call even when the first call is a long route that takes substantially longer to execute.

> there was a jobController in MapboxRouteLineView, it was used to make the call non-blocking

There's no mutable state in MapboxRouteLineView and all of the maps calls were wrapped in a single coroutine on the main thread. In addition calls to the MapboxRouteLineView render method are on the main thread.  I don't think there's any real risk of of a race condition but I've left the coroutine out of MapboxRouteLineView for the time being. It only saves a few milliseconds anyway.

korshaknn commented:
@cafesilencio 
> All of the async calls run on the main thread with a thread from the IO pool being used for specific data calculation tasks which don't mutate any state.

now each method call has the next block:
```kotlin
jobControl.scope.launch {
            mutex.withLock {
                    // some logic
            }
}
```
docs says about the `mutex`:
` It is **non-reentrant**, that is invoking [lock] even from the same thread/coroutine that currently holds
 the lock still suspends the invoker.` 

we use `main` thread to launch each coroutine, so each coroutine will be suspended until previous one is finished.
if we have some long-time operation inside `mutex`:
`
mutex.withLock {
                    withContext(Dispatcher.IO){
                    // long-time background work
                    }
 }`
we will be suspended until it's is finished even it runs on worker thread. Because we wrap with `mutex` all code inside `launch`.
 
do we want such behaviour? 
maybe it's wrong and mutex should be used as less as possible? we should wrap with it critical places where we change mutable state only.

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt:
```diff
@@ -168,42 +173,66 @@ class MapboxRouteLineApi(
+                    it.add(0, route)
+                }
+                val featureDataProvider: () -> List<RouteFeatureData> =
+                    MapboxRouteLineUtils.getRouteFeatureDataProvider(newRoutes)
```
Why do we re-decode the geometry from polyline to feature collection whenever an alternative is selected? The geometries don't change, we should be able to cache them. This looks to be one of the slowdowns in the alternative route line selection.

We should also execute it on a worker thread instead of the main thread, but that wouldn't resolve the unnecessary duplicate caluclations.

LukasPaczos replied:
We can treat this as a follow-up.

cafesilencio replied:
For clarification MapboxRouteLineUtils.getRouteFeatureDataProvider(newRoutes) returns a function. The function is eventually invoked on a background thread in buildDrawRouteState().

```
        val routeFeatureDataDef = jobControl.scope.async(ThreadController.IODispatcher) {
            featureDataProvider()
        }
```

LukasPaczos commented on libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt:
```diff
@@ -168,42 +173,66 @@ class MapboxRouteLineApi(
+        jobControl.scope.launch {
+            mutex.withLock {
+                val featureDataProvider: () -> List<RouteFeatureData> =
+                    MapboxRouteLineUtils.getRouteFeatureDataProvider(directionsRoutes)
```
Same as above.

cafesilencio replied:
Good catch, I'll cut a ticket.

LukasPaczos reviewed (commented):
> we use main thread to launch each coroutine, so each coroutine will be suspended until previous one is finished

I feel like we could merge the current setup, and iterate. The current PR does not support coalescing or cancellation at all as you mentioned, there's always an ordered queue waiting for execution. This successfully offloads the work to the worker thread which is a great first step, and in a follow-up PR we can introduce cancelation policies.

I have a couple of smaller comments below, but other than that the PR LGTM.

LukasPaczos commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/SetRouteCancellationTest.kt:
```diff
@@ -0,0 +1,201 @@
+import java.util.Timer
+import kotlin.concurrent.timerTask
+
+class SetRouteCancellationTest : BaseTest<BasicNavigationViewActivity>(
```
Yea, we shouldn't need to wait for the first job to finish if we already know that it's going to be overridden immediately.

Could you cut a follow-up ticket once this lands?

cafesilencio replied:
The documentation described another strategy using an actor but that was more refactoring than I wanted to do at this moment especially considering the route line api has been stable. It's worth exploring other strategies in the near future.

LukasPaczos replied:
Could you also cut a follow up ticket for this one?

LukasPaczos replied:
https://github.com/mapbox/mapbox-navigation-android/issues/4232

cafesilencio commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/SetRouteCancellationTest.kt:
```diff
@@ -0,0 +1,201 @@
+            timerTask {
+                routeLineApi.setRoutes(shortRoutes, consumer)
+            },
+            145
```
I experimented with different values. On my device 145 seemed to be a sweet spot but of course results will vary from device to device. I didn't want to make the time too short but I'll adjust it.

LukasPaczos replied:
Have you seen the test behaving differently when you decreased the delay time?

Reading from coroutine `launch` docs, the job should be scheduled for execution immediately, so even no delay should result in the behavior that we're trying to test, but I guess we can have a small delay to actually try to mimic an async behavior.

cafesilencio commented on instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/ui/routeline/SetRouteCancellationTest.kt:
```diff
@@ -0,0 +1,201 @@
+        val routeLines = listOf(RouteLine(route, null))
+
+        val consumer = object : MapboxNavigationConsumer<Expected<RouteSetValue, RouteLineError>> {
+            override fun accept(value: Expected<RouteSetValue, RouteLineError>) {
+                val primaryRoute = routeLineApi.getPrimaryRoute()
```
The ordering is in the other test where the same consumer is used.  This test was intended to prove each consumer gets the right result and the state of the routelineApi class is correct from the consumer's point of view. It was meant to test that two different calls with their own consumers made (almost) simultaneously would get the right result. Specifically there is a check for routeLineApi. getPrimaryRoute() indicating the state was correct.

LukasPaczos replied:
The other test verifies the response order of 2 `setRoute` calls, not the order of `setRoute` and `clearRoute` which also needs to be verified.

cafesilencio replied:
Deleted my previous comment. I thought your comment was referring to a different test. I didn't read it closely enough.

LukasPaczos reviewed (approved):
LGTM, thanks for addressing the comments @cafesilencio!

LukasPaczos commented:
@cafesilencio could you also update the migration guide?

cafesilencio commented:
Updated
